### PR TITLE
`Programming exercises`: Fix broken download button for auxiliary repositories

### PR DESCRIPTION
**Pull request by iyannsch**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

Issue #8605 describes the current state in which the download of non-existent auxiliary repositories fail silently. This is not a problem for building or testing programming exercises, but heavily limit the usability for novice users. 

When auxiliary repos are created during exercise creation, the auxiliary repos are created and linked correctly. In this case, the current implementation successfully provided a download.
However, when aux. repos are added to existing exercises, no directory is created in LocalVC, but the DB is kept in sync and represents the changes. The instructor performing the action is informed about this situation but no fix is suggested nor is a permanent error message raised. When trying to clone or download the non-existent repository, an internal server error (500) arises but no information is shown in the UI. This leaves teams of multiple instructors with the situation that they might not understand why an aux. repo does not work.

### Description

I propose a more sophisticated solution changing the way how auxiliary repositories are added to already existing exercises.
Adding new aux repos to existing exercises (and a set of already existing aux repos), poses the challenge of not overwriting those aux repos that are already referenced in tests etc. 
[This code](https://github.com/ls1intum/Artemis/blob/1348a86de1ce91d19422687567a7964fc9d9cc95/src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts#L631-L636) contains the logic for detecting "illegal" auxiliary repo changes. As soon as a `name` or `checkoutDirectory` are changed, it is considered illegal. 

As a first step, I implemented an error handler for failing repository downloads. Users are not informed about the situation and know that their repository does not exist. This fights the symptoms but not the root-cause. 

To solve the underlying problem, I added the functionality in the server to keep the filesystem in sync with the database when deleting and creating auxiliary repositories. The instructor can add and remove auxiliary repositories during both exercise creation and modification. Editing existing auxiliary repositories (e.g., changing the repo's name) is not supported and will still throw the error.

### Discussion for further improvement
Currently, the `VersionControlService`-Interface does not offer the functionality to modify existing repositories. Systems like GitLab that were used in the past do not offer such functionality with Artemis as well. It could make sense to extend the LocalVC with editing functionality for auxiliary repos - currently, I'll prioritize more pressing issues and leave this for future work.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course

#### Verify successful download on creation
1. Log in to Artemis
2. Create a new programming exercise and add an auxiliary repository before saving the exercise
3. Verify that you can download the aux repo 
image

4. Verify that the banner is shown
image

#### Verify successful download on modification
1. Log in to Artemis
2. Edit an existing programming exercise and add an auxiliary repository before saving the exercise
3. Verify that you can download the aux repo 
image

4. Verify that the banner is shown
image

#### Verify error alert when misconfiguring repositories 
1. Log in to Artemis
2. Edit an existing programming exercise
3. Rename an existing auxiliary repository and save
4. Verify that the banner is shown 
image

#### Verify error handling on misconfigured repositories
1. Log in to Artemis
2. Edit an existing programming exercise
3. Rename an existing auxiliary repository and save
4. Click on the `Download Repository` button and verify that you see an error banner 
image

#### Exam Mode Testing

<same as above>

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- **Style**
  - Improved the layout of the detail overview list by adding top margin to list items.

- **Bug Fixes**
  - Enhanced error handling in repository export functionality for programming exercises, providing better feedback when an export fails.

- **Documentation**
  - Added error messages for failed repository exports in both German and English to improve user feedback during export processes.

### CODE REVIEW

File: src/main/webapp/app/exercises/programming/shared/actions/programming-exercise-instructor-repo-download.component.ts [hunk 1]
```diff
@@ -6,6 +6,9 @@ import { downloadZipFileFromResponse } from 'app/shared/util/download.util';
 import { AlertService } from 'app/core/util/alert.service';
 import { faDownload } from '@fortawesome/free-solid-svg-icons';
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/programming/shared/actions/programming-exercise-instructor-repo-download.component.ts:8]:
```suggestion
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Regarding the root cause:
With external VCS like Bitbucket or Gitlab it was hard to track what users do without the knowledge of Artemis (creating additional repositories, changing build plans, ...). So it was hard for Artemis to track these changes accurately. Additionally it was hard to adapt the potentially custom build plans to correctly checkout newly added auxiliary repositories later on.

With LocalVC these problems should no longer occur, so we can provide instructors with the option to manage auxiliary repositories right in Artemis and then make sure that no additional repositories exist that are not accounted for or that auxiliary repositories exist in the Artemis database but not the LocalVC

File: src/main/webapp/app/exercises/programming/shared/actions/programming-exercise-instructor-repo-download.component.ts [hunk 2]
```diff
@@ -40,10 +43,23 @@ export class ProgrammingExerciseInstructorRepoDownloadComponent {
...
+                        // Return an observable of undefined so the subscribe method can continue
+                        return of(undefined);
+                    }),
+                )
+                .subscribe((response) => {
+                    if (response !== undefined) {
+                        downloadZipFileFromResponse(response);
+                        this.alertService.success('artemisApp.programmingExercise.export.successMessageRepo');
+                    }
+                });
```
[the remark this answers is not in the document]
Comment (iyannsch) [reply src/main/webapp/app/exercises/programming/shared/actions/programming-exercise-instructor-repo-download.component.ts:62]: it is

**Comment by iyannsch** [comment]
Don´t close it. @krusche is this a relevant feature to implement as @JohannesStoehr suggested?

**Comment by krusche** [comment]
I suggest we fix the root cause and allow more options for adding / editing / removing aux repos in LocalVC

**Comment by krusche** [comment]
Ideally, we provide more options in a follow-up and fix the root cause / bug first quickly, this PR is already open too long

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseRepositoryService.java [hunk 3]
```diff
@@ -229,6 +229,13 @@ void createRepositoriesForNewExercise(final ProgrammingExercise programmingExerc
...
     }
 
+    /**
+     * Creates and initializes all auxiliary repositories for a new programming exercise.
+     *
+     * @param projectKey          The key of the project the exercise belongs to. Can be found in the programming exercise.
+     * @param programmingExercise The programming exercise for which the auxiliary repositories should be created.
+     * @throws GitAPIException Thrown in case creating a repository fails.
+     */
     private void createAndInitializeAuxiliaryRepositories(final String projectKey, final ProgrammingExercise programmingExercise) throws GitAPIException {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseRepositoryService.java:239]: For a follow up:
If we already pass the programming exercise that contains the project key, why do we need to pass the key separately as well?

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts [hunk 4]
```diff
@@ -627,14 +637,51 @@ export class ProgrammingExerciseUpdateComponent implements AfterViewInit, OnDest
...
         }
 
-        // If the auxiliary repositories were edited after the creation of the exercise, the changes have to be done manually in the VCS and CIS
+        /*
+         If properties for an auxiliary repository were edited, the changes have to be done manually in the VCS and CIS.
+         Creating or deleting new auxiliary repositories works automatically and does not throw a warning.
+         We check that by verifying all "current" repositories with an ID (meaning they are not new) with their backup.
+         */
+        //
+
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:646]:
```suggestion
```

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts [hunk 5]
```diff
@@ -627,14 +637,51 @@ export class ProgrammingExerciseUpdateComponent implements AfterViewInit, OnDest
...
+                        ' and checkoutDirectory ' +
+                        auxRepo.checkoutDirectory +
+                        ' with ' +
+                        backupAuxRepo.checkoutDirectory,
+                );
+
+                console.log(
+                    'Result: ' + (backupAuxRepo.id === auxRepo.id && (backupAuxRepo.name !== auxRepo.name || backupAuxRepo.checkoutDirectory !== auxRepo.checkoutDirectory)),
+                );
+
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:671]:
```suggestion
```

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts [hunk 6]
```diff
@@ -627,14 +637,51 @@ export class ProgrammingExerciseUpdateComponent implements AfterViewInit, OnDest
...
+        /*
         const changedAuxiliaryRepository =
             (this.programmingExercise.auxiliaryRepositories?.length ?? 0) < (this.backupExercise?.auxiliaryRepositories?.length ?? 0) ||
             this.programmingExercise.auxiliaryRepositories?.some((auxRepo, index) => {
                 const otherAuxRepo = this.backupExercise?.auxiliaryRepositories?.[index];
                 return !otherAuxRepo || auxRepo.name !== otherAuxRepo.name || auxRepo.checkoutDirectory !== otherAuxRepo.checkoutDirectory;
             });
-        if (changedAuxiliaryRepository && this.programmingExercise.id) {
+         */
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:683]: Please remove the uncommented code if it is no longer needed

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseRepositoryService.java (hunk 3 above)
Comment (iyannsch) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseRepositoryService.java:239]: I think that the auxiliary repositories are generated before the exercise is saved (if I remember correctly)

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by marlon-luca-bu — APPROVED** [review]
Code

**Review by BBesrour — APPROVED** [review]
Tested on TS1. works as described.

Should the warning be displayed when changing the checkoutPath?

**Comment by iyannsch** [comment]
@BBesrour I think for more complicated Build Jobs this could make sense. Regardless of this, it was initially implemented like this - I only removed the alert for creation and deletion 👍

**Pull request merged**